errors/errors.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errors
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+	unknown                      = "UNKNOWN_REASON"
+)
+
+const (
+	// gRPC to HTTP Mapping: 500 Internal Server Error
+	unknownHTTPCode = http.StatusInternalServerError
+)
+
+var UnknownErrorReason = WithErrorReason(unknown, unknownHTTPCode, codes.Unknown)
+
+type ResourceInfo struct {
+	Type string
+	Name string
+}
+
+// call this function to apply option
+type Option func(*Error)
+
+type Error struct {
+	err            error
+	description    string
+	reason         string
+	httpCode       int
+	grpcStatusCode codes.Code
+	metadata       map[string]string
+	resourceInfo   *ResourceInfo
+}
+
+// EnableComponentErrorCode stores an indicator for
+// components to determine if the ErrorCodes Feature
+// is enable.
+func EnableComponentErrorCode(metadata map[string]string) {
+	if metadata != nil {
+		metadata[ErrorCodesFeatureMetadataKey] = "true"
+	}
+}
+
+func featureEnabled(metadata map[string]string) bool {
+	_, ok := metadata[ErrorCodesFeatureMetadataKey]
+	return ok
+}
+
+// New create a new Error using the supplied metadata and ErrorOptions
+// **Note**: As this code is in `Feature Preview`, it will only continue processing
+// if the ErrorCodes is enabled
+// TODO: @robertojrojas update when feature is ready.
+func New(err error, metadata map[string]string, options ...Option) *Error {
+	// The following condition can be removed once the
+	// Error Codes Feature is GA
+	if !featureEnabled(metadata) {
+		return nil
+	}
```

Comment:
Should this no be returning the `err` as is? This seems like a code smell that the caller is going to have to learn the feature gate anyway.